### PR TITLE
Make fields orderable

### LINKED ISSUE #3720: Allow ordering of columns in Data Model
Currently it is not possible to change ordering of columns in Data Model tab. It seems that columns are ordered by database column name. If the business name is different to database column name, it is hard to find requested field.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by sbelak**
Makes fields orderable. Introduces 4 ordering strategies: 
* `:database`     - use the same order as in the table definition in the DB;
* `:alphabetical` - order alphabetically by name;
* `:custom`       - the user manually set the order in the data model
* `:smart`        - Try to be smart and order like you'd usually want it: first PK, followed by `:type/Name`s, then `:type/Temporal`s, and from there on in alphabetical order.

Default is `:database`. 

The setting is stored per-table in `:field_order` field (and can be updated via `PUT /api/table/:id` as usual).

Fields should be ordered using `:position` field (the backend also always sorts the fields in the response according to that).

## API endpoint for custom order
`PUT /api/table/:id/fields/order` with body being a vector of field IDs as they should be ordered. 

## Note

For Mongo `database` order doesn't really make sense and Druid's API returns fields as a map, so ordering is not guaranteed to be preserved (depends on the number of fields and Clojure internals). Should those databases default to `:smart`? 

~The `:position` field is only used for custom order and does not get overwritten if you eg. change order to `:database`. There is no benefit to caching the current sort order and this way you don't loose what you handcrafted if you mess around which I think is a nice property.~

`:database_position` will get populated during first/next sync.

Closes #3720

### CODE REVIEW

File: src/metabase/models/table.clj
Comment (sbelak) [inline src/metabase/models/table.clj:190]: This is not ideal. Currently `with-fields` is only used in transforms where we don't care about field order, so I don't think it's worth the hoop jumping this would require to be both efficient and respect `:field_order`.

**Comment by paulrosenzweig** [comment]
I added a FE for the datamodel pages:

column-ordering

@sbelak, how can I display the fields correctly for non-custom? It seems like the `position` is always set for the last custom order, is that right? Could the fields come down sorted according to the current ordering strategy? I guess hidden fields could be sorted last.

**Comment by sbelak** [comment]
@paulrosenzweig do you need explicit ordering information? The way I did it is that the order in which fields come back is what it's suppose to be. But looking at your gif, I'm thinking this is probably not the best approach.

**Comment by mazameli** [comment]
> Thoughts @mazameli

I'm not understanding the question, @sbelak. My expectations are:

- The columns should be listed in the order of the chosen sort order. If I have `Database` order selected, the first column listed should be whatever's first in the database (probably something like `ID`).
- Selecting a different sort order should immediately change the listed order of the columns. If I had previously had `Database` selected and then select `Alphabetical`, I should see e.g. `Address` as the first column instead of `ID`.

Also, @paulrosenzweig, I'd like for the draggable handles to always be present and usable. If I have e.g. `Database` order selected and I think drag and reorder a column, the search order dropdown should automatically change to `Custom`, overwriting any previously stored custom order.

Happy to talk through this "in person" if that doesn't clear things up.

**Comment by sbelak** [comment]
@paulrosenzweig what would be a useful response from /fields/order endpoint?

**Comment by sbelak** [comment]
How shall we handle custom sorting in the presence of hidden fields. If we have hidden fields, the field list in Data Model is only partial, what position should hidden fields then get? At the end?

**Comment by mazameli** [comment]
I just pushed a change to get the gear button and description input to line up more nicely like in my mocks. Screenshot:

image

There are two stylistic things I couldn't tweak, though:

The Visibility and Type selects seem to have gotten taller in this PR vs. master, making them look weird next to the Name input box. They should be the same height as the input box:

image

The dropshadows on the column items shouldn't be there unless an item is being dragged to reorder it. This looks too busy:

image

Here's the desired design, which I tried to implement by switching out the `Card` component for a simple div, but that seemed to break the dragging interaction:

image

**Comment by sbelak** [comment]
Change the response that it now
1) `:position` is always updated to the current ordering and can be used for sorting
2) `PUT /api/table/:id` returns hydrated fields if `field_order` has changed

**Comment by mazameli** [comment]
Looks like column order isn't being adhered to in the Data Reference.

### Data Reference fields page for a given table:

image

### Data model order:

image

**Comment by paulrosenzweig** [comment]
@mazameli good catch! I had forgotten about the data refs.

**Comment by mazameli** [comment]
This behavior doesn't look right: if I have a custom order, and I then choose e.g. Database order, if I then manually drag and reorder a column, all the columns revert to the _previous_ custom ordering. In this situation I expected my old custom ordering to be forgotten, and for this to become my new custom ordering. 

(If I had used the dropdown to switch from `Database` to `Custom`, without dragging columns around, I would then have expected my previous custom ordering to have been restored.)

weird-custom-order

**Review by mazameli — APPROVED** [review]
UX and UI look good to me! I checked everywhere in the product I could think of where the column order matters and it matched the column ordering I chose (the query builder, viz settings, the data reference, field popover selects). 

The one place where we always use alphabetical order is for the autocomplete menu when writing custom expressions. I think this actually feels right, since that list updates as you type; the alpha order seems expected to me.

**Comment by sbelak** [comment]
@mazameli for DBs where DB order doesn't make sense (or we can't get it due to driver limitations), should we default to something else or just leave it as is?

**Comment by mazameli** [comment]
> for DBs where DB order doesn't make sense

Can you give me an example where DB order doesn't make sense?

> should we default to something else or just leave it as is?

What would "as is" mean here? Are you talking about already-connected databases?

**Comment by sbelak** [comment]
Document databases (ie. Mongo) don't map very cleanly to our ordered world view, especially when you have nesting and heterogeneous documents. Additionally the Druid driver we're using can switch the order around for tables with > 16 fields (16 is a current Clojure implementation specific and can change in future releases). 

As is was meant as it's currently with this PR, so do something resembling DB order, but not be too bothered if it's not perfect.

**Comment by mazameli** [comment]
> As is was meant as it's currently with this PR, so do something resembling DB order, but not be too bothered if it's not perfect.

Gotcha, yeah, in these cases it seems like doing something resembling DB is probably better, or at least no worse than alpha order. I'm not hung up on this.

**Comment by mazameli** [comment]
From the demo today:
- When @paulrosenzweig tried to apply `Database` ordering the the `Orders` table, it appeared like nothing happened.
- We should verify that tables in already-connected DBs will switch over to the `Database` ordering, but that saved questions aren't altered or broken.

**Comment by paulrosenzweig** [comment]
@sbelak, the source of the bug we saw yesterday was that sometimes each field's position was set to `0`. It looks like newly synced fields correctly set `position` to `database-position`, so I guess maybe this is just an issue for tables that existed before upgrading to this code.

I couldn't figure out where to fix it correctly in the backend. Instead, I worked around the issue in the frontend by using the implicit field order if each field has `position: 0`. That fixes the issue we saw, but I don't know if there are any other bugs that would crop up if we leave these positions uninitialized.

---

@mazameli, I checked how saved questions are handled when upgrading to this. If a question selects all columns from a table and doesn't reorder them, the order of its fields will change. If the columns were edited at all, the order is preserved through the upgrade.

While technically that does mean some saved questions will change, I think that's ok. Those questions are effectively doing a `select *`. It feels like the user didn't really express any preference in columns and just said "give me everything".

**Comment by mazameli** [comment]
@paulrosenzweig Cool, I think that's acceptable re: saved questions. Does it maintain column order only if they've reordered columns, or does it do that whenever they've set _any_ table viz settings at all (e.g. conditional formatting, date formatting, etc.)?

**Comment by paulrosenzweig** [comment]
@mazameli yeah the order is only locked if they've reordered, added, or removed columns. Those operations update the query with the desired fields in a set order. Adding formatting on a specific column doesn't affect the query itself but instead stores the setting keyed by column name on the card.

**Comment by sbelak** [comment]
@paulrosenzweig a sync needs to run for the fields to update ... which should happen if you upgrade and restart. Did you just switched branches yesterday with an already running instance?

**Comment by paulrosenzweig** [comment]
@sbelak I did that yesterday when switching away from `orderable-fields` I think. Today I repro'd starting on master with a fresh db then switching to this db and restarting the server.

**Comment by mazameli** [comment]
> That fixes the issue we saw, but I don't know if there are any other bugs that would crop up if we leave these positions uninitialized.

So it sounds like we're holding off on merging this so that we can figure out where and how to fix this more thoroughly on the backend?

**Comment by paulrosenzweig** [comment]
@sbelak I just repro'd again with this procedure:

I started on master with a fresh db. After completing the signup flow, I switch to this branch and restart the server. I triggered a sync manually, but after that completes `position` is still set to zero for all field in the sample db. Switching the table field order from "database" to "custom" sets `position` on each field.

**Comment by sbelak** [comment]
I was wrong, we don't run sync automatically on upgrade and @camsaul has kindly reminded me that sync can take hours so we probably don't want to do that (and some people have it off). This makes the whole plan of switching to db ordering by default somewhat tricky. 

Options:
1) take advantage that in most cases field IDs are already in DB order and write a migration that prepopulates the `database_position` values using that (the correct values will then be set on first actual sync).
2) run sync in the background as we start this one time (but what about people who have it off?)
3) prompt the user to trigger sync if they want this
4) document this as applying only to databases added henceforth

My vote is 1)

Thoughts @mazameli, @camsaul?

**Review by camsaul — CHANGES_REQUESTED** [review]
Need to fix failing driver tests before merging. I'm working on this right now

**Comment by camsaul** [comment]
@sb I fixed the tests for BigQuery, Druid, and Snowflake. Those we the runs reusing cached data. There are still a handful of driver tests left to fix.

**Comment by sbelak** [comment]
> I was wrong, we don't run sync automatically on upgrade and @camsaul has kindly reminded me that sync can take hours so we probably don't want to do that (and some people have it off). This makes the whole plan of switching to db ordering by default somewhat tricky.
> 
> Options:
> 
> 1. take advantage that in most cases field IDs are already in DB order and write a migration that prepopulates the `database_position` values using that (the correct values will then be set on first actual sync).
> 2. run sync in the background as we start this one time (but what about people who have it off?)
> 3. prompt the user to trigger sync if they want this
> 4. document this as applying only to databases added henceforth
> 
> My vote is 1)
> 
> Thoughts @mazameli, @camsaul?

bump

**Review by camsaul — APPROVED** [review]

File: src/metabase/models/table.clj
Comment (sbelak) [inline src/metabase/models/table.clj:42]: what's our convention here? Do we simply rely on the DB default or do we set them on the model level as well?

**Comment by mazameli** [comment]
@sbelak I think what you outlined in option 1 sounds perfectly acceptable.

File: src/metabase/models/table.clj
Comment (camsaul) [reply src/metabase/models/table.clj:42]: it's best to set the default at the DB level if possible. If it's set there than there's no real reason to set it in the model level

**Comment by sbelak** [comment]
I've added the seeding migration and made default order be a per-driver setting. Doing per-db will I assume require some design work (do we really want to cram it into the DB adding page), so it's perhaps better to tackle as a followup pr.

**Review by camsaul — APPROVED** [review]
the migration change works for me. Everything else I already looked at and it still looks good. Let's merge this

**Pull request merged**
